Add right panel with Room information

MidhunSureshR opened the pull request.

MidhunSureshR committed:
Add view and vm for RoomInformation

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Allow details to be child of room

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add code to toggle RoomInformation panel

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Create RoomInformationView on toggle

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Put name of property in emitChange

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Handle case when path.get() may be null

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Allow details to be child of rooms

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove right panel on grid update if needed

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Dispose vm preemptively

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Extract method

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove unnecessary check

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add CSS to display sidebar to the right

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fallback to canonical alias if no names found

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add background for sidebar

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add room avatar to RoomInfoView

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make room avatar larger in RoomInfoView

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make display name bold

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Better styling

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Create method to add sidebar element

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add method to create avatar section

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Bring in icon assets

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add avatar size for design

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add css to place items appropriately

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add styling for RoomInfoView

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add close button to RoomInfoView

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Position the close button to right end

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Style close button

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make close button close the view

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Bring in info icon

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add button in Room header to open details

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Style open button

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Show panel when first visit contains /details

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make side-bar responsive

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Justify content to flex-end instead

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Info icon should open/close view alternatively

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Prefer canonical alias over room_id if available

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -285,6 +289,13 @@ export class RoomViewModel extends ViewModel {
+        const isPanelOpen = !!this.navigation.path.get("details");
+        const link = isPanelOpen ? this.urlCreator.urlUntilSegment("room"):
+        this.urlCreator.urlForSegment("details");
+        window.location.href = link;
```
view models should not access any DOM interfaces, like window, as they are assume to work outside of browser environments. The easiest way would probably to expose a url property from this view model.

bwindels commented on src/domain/session/SessionViewModel.js:
```diff
@@ -112,9 +118,15 @@ export class SessionViewModel extends ViewModel {
         const changed = !(this._gridViewModel && roomIds);
         const currentRoomId = this.navigation.path.get("room");
+        // Close right-panel if needed
+        this._toggleRoomInformationPanel();
```
you should not have to react to a change in the grid rooms set, as the active room will be a room segment after that. So you can just react to a change in the active room, and it will work for both grid and single rooms.

So, concretely, you can call _toggleRoomInformationPanel from `this.track(currentRoomId.subscribe(roomId => ` on line 62, and in your toggle method check if the room id of the current vm, if any is still the active one.

MidhunSureshR committed:
Keep RoomInfoView open across room/grid changes

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add link support to menu

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add menu entry to launch RoomInfoView

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Style links and buttons similarly in menu

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make font-size equal to that of room-header

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Put sidebar rows into container

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Do not show room id instead of canonical alias

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/platform/web/ui/session/room/RoomView.js:
```diff
@@ -69,14 +69,15 @@ export class RoomView extends TemplateView {
-                options.push(Menu.option(vm.i18n`Rejoin room`, () => vm.rejoinRoom()));
+                options.push(Menu.optionWithButton(vm.i18n`Rejoin room`, () => vm.rejoinRoom()));
             }
+            options.push(Menu.optionWithLink(vm.i18n`Room Details`, vm.detailsLink))
```
The font-size (and font-family of the button, but that should be fixed upstream) seems off still for link menu options:
![image](https://user-images.githubusercontent.com/274386/120608280-8a2ba800-c440-11eb-9ae5-89c7b705d68a.png)

bwindels replied:
Also, perhaps put this option first in the menu, so you're less likely to hit (the destructive) "Leave room" option by accident.

bwindels commented on src/platform/web/ui/session/rightpanel/RoomInfoView.js:
```diff
@@ -0,0 +1,47 @@
+export class RoomInfoView extends TemplateView {
+
+    render(t, vm) {
+        const encryptionString = vm.isEncrypted ? "On" : "Off";
```
For all user visible strings, please use
```js
vm.i18n`...`
```
which should make it easier to introduce translations once we decide to support them.

bwindels commented on src/domain/navigation/index.js:
```diff
@@ -35,9 +35,9 @@ function allowsChild(parent, child) {
         case "rooms":
             // downside of the approach: both of these will control which tile is selected
-            return type === "room" || type === "empty-grid-tile";
+            return type === "room" || type === "empty-grid-tile" || type === "details";
```
Is this needed?
A `details` segment would always only come after a `room` segment, no?
And room can occur either after session or rooms.
So you'd have:
`session > rooms > room > details` for the right panel in the grid view or
`session > room > details` for the right panel in the single room view.

Possible I'm missing something though.

bwindels commented on src/domain/session/RoomGridViewModel.js:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 */
 
-import {ViewModel} from "../ViewModel.js";
+import { ViewModel } from "../ViewModel.js";
```
I imagine your editor is doing this? The (very limited) code guidelines don't say anything either way in this regard, but most places in the project don't put spaces there and it's perhaps best not to clutter the PR with unrelated stylistic changes like this.

bwindels commented on src/domain/session/RoomGridViewModel.js:
```diff
@@ -84,7 +84,11 @@ export class RoomGridViewModel extends ViewModel {
         }
         const vmo = this._viewModelsObservables[index];
         if (vmo) {
+            const detailsShown = !!this.navigation.path.get("details");
```
Also check the value of the segment here, like you do in parseUrlPath

bwindels commented on src/platform/web/ui/session/SessionView.js:
```diff
@@ -25,13 +25,15 @@ import {StaticView} from "../general/StaticView.js";
                 "SessionView": true,
-                "middle-shown": vm => !!vm.activeMiddleViewModel
+                "middle-shown": vm => !!vm.activeMiddleViewModel,
+                "SideBarActive": vm => !!vm.roomInfoViewModel
```
Granted that the css naming conventions are wildly inconsistent, but perhaps name this similar to the class for the middle area, e.g. `right-shown`?

bwindels commented on src/platform/web/ui/general/Menu.js:
```diff
@@ -15,33 +15,58 @@ limitations under the License.
     }
 
+    setCallback(callback) {
+        this.callback = callback;
```
Consider return `this` like in `setIcon` here so it can be used in a builder-like pattern: `new MenuOption().setCallback().setIcon()` ...

bwindels commented on src/domain/navigation/index.js:
```diff
@@ -113,6 +113,9 @@ export function parseUrlPath(urlPath, currentNavPath, defaultSessionId) {
                 segments.push(roomsSegmentWithRoom(rooms, roomId, currentNavPath));
             }
             segments.push(new Segment("room", roomId));
+            if (currentNavPath.get("details")?.value) {
```
Would you consider writing a small test below for this behaviour? :pray:

bwindels commented on src/domain/session/RoomGridViewModel.js:
```diff
@@ -196,7 +200,7 @@ export function tests() {
 
     class RoomViewModelObservableMock extends ObservableValue {
-        async initialize() {}
+        async initialize() { }
```
Same, please avoid changes like this

bwindels commented on src/domain/session/leftpanel/LeftPanelViewModel.js:
```diff
@@ -93,11 +93,13 @@ export class LeftPanelViewModel extends ViewModel {
             const room = this.navigation.path.get("room");
             if (room) {
                 path = path.with(room);
+                path = path.with(details);
```
This will throw if details is not present, so need to check here first.

bwindels commented on src/domain/session/leftpanel/LeftPanelViewModel.js:
```diff
@@ -106,6 +108,7 @@ export class LeftPanelViewModel extends ViewModel {
             if (room) {
                 path = path.with(this.navigation.segment("rooms", [room.value]));
                 path = path.with(room);
+                path = path.with(details);
```
Same as above

bwindels commented on src/domain/session/rightpanel/RoomInfoViewModel.js:
```diff
@@ -0,0 +1,50 @@
+    constructor(options) {
+        super(options);
+        this._room = options.room;
+        this._roomSummary = this._room._summary._data;
```
wrt to `_summary`, properties that start with an `_` are considered private, and it's a no-no to access them from anywhere but `this` (you might see it happen in tests sometimes, but that's probably being lazy). See below how to get access to the values you need.

bwindels commented on src/domain/session/rightpanel/RoomInfoViewModel.js:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    get isEncrypted() {
+        return !!this._roomSummary.encryption;
```
you can use `this._room.isEncrypted` here, see `BaseRoom` again.

bwindels commented on src/domain/session/rightpanel/RoomInfoViewModel.js:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    get name() {
+        return this._roomSummary.name || this._room._heroes?._roomName || this._roomSummary.canonicalAlias;
```
you don't need to do this for the name, `this._room.name` (see `BaseRoom.name`) will give you the correct name, taking into account heroes and canonical alias.

bwindels commented on src/domain/session/rightpanel/RoomInfoViewModel.js:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    get canonicalAlias() {
+        return this._roomSummary.canonicalAlias;
```
Please create a getter in BaseRoom for it.

bwindels commented on src/domain/session/rightpanel/RoomInfoViewModel.js:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    get memberCount() {
+        return this._roomSummary.joinCount;
```
Please create a getter in BaseRoom for it, perhaps `joinedMemberCount`?

bwindels commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -106,6 +106,10 @@ export class RoomViewModel extends ViewModel {
     get isEncrypted() { return this._room.isEncrypted; }
 
+    get roomDetailsLink() {
+        return this.urlCreator.urlForSegment("details");
```
Seems like this is not used?

bwindels commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -285,7 +289,11 @@ export class RoomViewModel extends ViewModel {
-    
+
+    get detailsLink() {
+        return this.urlCreator.urlForSegment("details");
```
Probably best to create the URLs in the constructor and store them in a property. The property on a view model should ideally be very fast (and not create new strings, etc), so the bindings reevaluating them won't slow things down.

bwindels commented on src/domain/session/rightpanel/RoomInfoViewModel.js:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    get closeLink() {
+        return this.urlCreator.urlUntilSegment("room");
```
Probably best to create the URLs in the constructor and store them in a property.

bwindels commented on src/domain/session/SessionViewModel.js:
```diff
@@ -78,6 +80,11 @@ export class SessionViewModel extends ViewModel {
+        const details = this.navigation.observe("details");
+        this.track(details.subscribe(() => this._toggleRoomInformationPanel()));
+        this._toggleRoomInformationPanel();
+
```
No need for this empty line at the end of the method probably?

bwindels commented on src/domain/session/SessionViewModel.js:
```diff
@@ -246,4 +256,21 @@ export class SessionViewModel extends ViewModel {
+    _toggleRoomInformationPanel() {
+        this._roomInfoViewModel = this.disposeTracked(this._roomInfoViewModel);
+        const room = this._roomFromNavigation();
+        const enable = !!this.navigation.path.get("details");
```
Like elsewhere, also check the value of the segment here, if present.

bwindels commented on src/domain/session/SessionViewModel.js:
```diff
@@ -246,4 +256,21 @@ export class SessionViewModel extends ViewModel {
+
+    _toggleRoomInformationPanel() {
+        this._roomInfoViewModel = this.disposeTracked(this._roomInfoViewModel);
+        const room = this._roomFromNavigation();
```
Do this inside the `if(enable)`?

bwindels commented on src/domain/navigation/index.js:
```diff
@@ -35,9 +35,9 @@ function allowsChild(parent, child) {
+            return type === "room" || type === "empty-grid-tile" || type === "details";
         case "room":
-            return type === "lightbox";
+            return type === "lightbox" || type === "details";
```
Hmm, just realized that opening the lightbox will close the details and vice-versa ... not sure how to solve this, but fine for now. Perhaps lightbox should not be part of the navigation path :thinking: as it is also annoying it ends up in the history, and is more of a modal thing. Anyway .... nothing for you to do here :smile:

bwindels commented on src/domain/session/RoomGridViewModel.js:
```diff
@@ -84,7 +84,11 @@ export class RoomGridViewModel extends ViewModel {
+            const detailsShown = !!this.navigation.path.get("details");
             this.navigation.push("room", vmo.id);
+            if (detailsShown) {
+                this.navigation.push("details", true);
```
Every push will cause the navigation to update. It's nice to make all the navigation changes in one go, so I'd suggest creating a new path (with room id and details) like already done in the left panel view model, and then applying it in one go, rather than multiple pushes.

bwindels commented on src/platform/web/ui/general/Menu.js:
```diff
@@ -15,33 +15,58 @@ limitations under the License.
 export class Menu extends TemplateView {
-    static option(label, callback) {
-        return new MenuOption(label, callback);
+    static optionWithButton(label, callback) {
```
This is already used in the composer (and since redactions is merged also in BaseMessageView), so either keep the name, or ensure you've applied the name change everywhere.

bwindels commented on src/domain/session/rightpanel/RoomInfoViewModel.js:
```diff
@@ -0,0 +1,50 @@
+export class RoomInfoViewModel extends ViewModel {
+    constructor(options) {
+        super(options);
+        this._room = options.room;
```
To receive updates on any properties of the room, you'll need to listen to its `change` event and call `emitChange`. Currently, if the room name, avatar or join count, ... would change, the right panel would not update. RoomViewModel does exactly the same, so serves as a good example.

bwindels commented on src/platform/web/ui/css/right-panel.css:
```diff
@@ -0,0 +1,31 @@
+    text-align: center;
+}
+
+.SidebarRow {
```
Were you envisioning that this be used in other sidebars in the app? I'd prefer calling anything related to the right panel that, rather than sidebar. This is consistent with left panel and is more specific which panel is meant.

bwindels commented on src/platform/web/ui/css/themes/element/theme.css:
```diff
@@ -43,14 +41,37 @@ limitations under the License.
-.hydrogen .avatar.usercolor6 { background-color: var(--usercolor6); }
-.hydrogen .avatar.usercolor7 { background-color: var(--usercolor7); }
-.hydrogen .avatar.usercolor8 { background-color: var(--usercolor8); }
+.hydrogen .avatar.usercolor1 {
```
I imagine your tooling got ahead of you here, but please don't make unrelated stylistic changes like this. For one, it makes it harder to see where you made actual changes in this file.

bwindels commented on src/platform/web/ui/css/themes/element/theme.css:
```diff
@@ -932,3 +978,87 @@ button.link {
+
+/* Right Panel */
+
+.RoomInfo {
```
So, as said, the CSS naming conventions are currently wildly inconsistent. But for CSS classes that are not used in multiple views (which I think applies to all the below?), I do try to use the convention of `ViewName` for the root element, and `ViewName_something` for sub elements.

bwindels replied:
As commented in RoomInfoViewModel, I'd suggest the name `RoomDetailsView` here.

I imagine once you add the member list to the right panel. RoomDetailsView would be one possible child of RightPanelView or something, with `MemberListView` being the other possibility.

bwindels commented on src/domain/session/rightpanel/RoomInfoViewModel.js:
```diff
@@ -0,0 +1,50 @@
+import { ViewModel } from "../../ViewModel.js";
+import { avatarInitials, getIdentifierColorNumber, getAvatarHttpUrl } from "../../avatar.js";
+
+export class RoomInfoViewModel extends ViewModel {
```
can we call this `RoomDetails{View|ViewModel}` everywhere? It's consistent with the details segment in the nav/url. Also the CSS classes, see comments in `element/theme.css`

bwindels commented on src/domain/session/SessionViewModel.js:
```diff
@@ -112,6 +119,10 @@ export class SessionViewModel extends ViewModel {
         return this._roomViewModelObservable?.get();
     }
 
+    get roomInfoViewModel() {
```
This would also become `roomDetailsViewModel` ...

bwindels reviewed (changes requested):
Thank you for your changed @MidhunSureshR! This is shaping up nicely! I did a full code review, don't be intimidated by the many comments, most should be very easy to address.

bwindels commented on src/platform/web/ui/css/themes/element/theme.css:
```diff
@@ -932,3 +978,87 @@ button.link {
+}
+
+.SidebarRow {
+    width: 90%;
```
Wonder what this is for? It is adding space on the right, making it look somewhat unbalanced:
![image](https://user-images.githubusercontent.com/274386/120623146-fdd4b180-c44e-11eb-92c8-945f390d0736.png)

bwindels commented on src/platform/web/ui/css/themes/element/theme.css:
```diff
@@ -932,3 +978,87 @@ button.link {
+
+.RoomInfo {
+    background: rgba(245, 245, 245, 0.90);
+    padding: 10px;
```
I'd go for 16px here. Our designer recommended me to always use values multiple of 4px in any case, so you end up with a 4px grid that looks more coherent.

bwindels commented on src/platform/web/ui/css/themes/element/theme.css:
```diff
@@ -932,3 +978,87 @@ button.link {
+}
+
+.RoomAvatar {
+    margin-top: 44px;
```
Hmm, not sure about this. I think I prefer how it looks without this (and the other changes I suggested):

![image](https://user-images.githubusercontent.com/274386/120623944-b69af080-c44f-11eb-93b9-e430afe22e5b.png)

bwindels reviewed (changes requested):
Had a look at the styling as well, mostly looks good, just a few suggestions.

MidhunSureshR committed:
Undo formatting/stylistic changes

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fix Menu: Use t instead of tag to get events

- Also move constructor up

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Use optionWithButton

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make menu-items uniform

- Same font ,font-size and height.

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Change Room Details --> Room details

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move Room details menu entry to top

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Use internationalization for user facing strings

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add missing checks

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Reduce duplication in code

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move details handling to separate function

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
details do not need to be child of rooms

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Calculate path when button is clicked

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Check value of details everywhere

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Change SideBarActive --> right-shown

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Modify setters in MenuOption to support chaining

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add new test for parseUrlPath

Make sure that the details panel stays open during room change.

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Do not use private props from room summary

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove unused getter

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Do not compute link in getter

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Use url instead of link

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove unncessary empty line

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move variable to scope of conditional

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove link support from Menu

- Not needed anymore since every link item has been rewritten as a
  button.

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Update on changes to vm

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Stick to naming convention for css/html classes

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Rename room info to room details everywhere

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove unwanted width

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make value multiple of 4

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove margin from top

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Refactor function

- Move related parameters closer together.
- Remove unused parameter.

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Increase height to make element more centered

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make formatting consistent

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add rule to warn on wrong formatting

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make formatting consistent with the codebase

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Undo spillover from earlier formatting havoc

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/platform/web/ui/session/rightpanel/RoomDetailsView.js:
```diff
@@ -0,0 +1,51 @@
+    _createRightPanelRow(t, label, labelClass, value) {
+        const labelClassString = classNames({RoomDetailsView_label: true, ...labelClass});
+        return t.div({className: "RoomDetailsView_row"}, [
+            t.div({className: labelClassString}, [text(label)]),
```
Just a small note that you don't need `text()` here. It's only when working with node bindings where you explicitly need to return a DOM node you need to do this. t.<element>() will convert any children strings to text nodes.

bwindels commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -283,10 +283,16 @@ export class RoomViewModel extends ViewModel {
         return this._composerVM;
     }
+
+    toggleDetailsPanel() {
```
the implementation looks like it will just open rather than toggle? Change name?

bwindels commented on src/domain/session/rightpanel/RoomDetailsViewModel.js:
```diff
@@ -0,0 +1,51 @@
+    constructor(options) {
+        super(options);
+        this._room = options.room;
+        this._room.on("change", () => this.emitChange());
```
You'll need to unsubscribe in the dispose method you override from ViewModel, which also means you can't use an anonymous lambda here. See RoomViewModel

bwindels reviewed (commented):
Thank you for addressing all the comments, just went over those and resolved accordingly. Came across one issue and a little remaining nit. I'll do another review of the whole thing (rather than just going over the comments as I did now), but I think we're mostly good to go.

MidhunSureshR committed:
Rename method

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove space after brace in two places

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Unsubscribe on dispose

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Undo lint config change

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/platform/web/ui/css/themes/element/theme.css:
```diff
@@ -660,14 +662,19 @@ button.link {
     background-color: transparent;
     text-align: left;
     padding: 8px 32px 8px 8px;
+    font-size: 1.6rem;
```
This should be 15px aka `1.5rem`

bwindels commented on src/platform/web/ui/css/layout.css:
```diff
@@ -65,7 +72,7 @@ the layout viewport up without resizing it when the keyboard shows */
 /* mobile layout */
 @media screen and (max-width: 800px) {
-    .SessionView:not(.middle-shown) {
+    .SessionView:not(.middle-shown, .right-shown) {
```
selector lists in `:not` are not supported in IE [apparently](https://developer.mozilla.org/en-US/docs/Web/CSS/:not). Can you write this out?

bwindels reviewed (changes requested):
Looking great, found a few more things, but after that we're good to merge I think.

MidhunSureshR committed:
Rename toggle --> open

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make navigation changes in one go

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove text(..)

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Change font size

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove selector list

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/domain/session/RoomGridViewModel.js:
```diff
@@ -183,6 +189,7 @@ export class RoomGridViewModel extends ViewModel {
 
 import {createNavigation} from "../navigation/index.js";
 import {ObservableValue} from "../../observable/ObservableValue.js";
+import { Segment } from "../navigation/Navigation.js";
```
Nit: only the imports needed for the tests should be below here, the others up at the top. You can also avoid the import by doing `this.navigation.segment(...)` fwiw.

bwindels commented on src/platform/web/ui/css/layout.css:
```diff
@@ -72,7 +72,7 @@ the layout viewport up without resizing it when the keyboard shows */
 /* mobile layout */
 @media screen and (max-width: 800px) {
-    .SessionView:not(.middle-shown, .right-shown) {
+    .SessionView:not(.middle-shown) {
```
is the `.right-shown` not needed then in here?

MidhunSureshR committed:
Remove Segment import

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove css assumption

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make sure room does exist before creating vm

- This will stop the code from throwing when opening /details on
  UnknownRoomView.

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels committed:
Merge branch 'master' into room-info
